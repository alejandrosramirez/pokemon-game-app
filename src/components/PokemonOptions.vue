<template>
	<div class="pokemon-options-container">
		<ul>
			<li
				v-for="pokemonOption in pokemonOptions"
				:key="pokemonOption.id"
				@click="$emit('selectedPokemon', pokemonOption.id)"
			>
				{{ pokemonOption.name }}
			</li>
		</ul>
	</div>
</template>

<script>
export default {
	name: "PokemonOptions",
	props: {
		pokemonOptions: {
			type: Array,
			required: true,
		},
	},
};
</script>

<style scoped>
.pokemon-options-container {
	display: flex;
	justify-content: center;
}
ul {
	list-style-type: none;
	margin: 20px 0 0 0;
	padding: 0;
}
li {
	background-color: white;
	border-radius: 5px;
	border: 1px solid rgba(0, 0, 0, 0.2);
	cursor: pointer;
	margin-bottom: 10px;
	width: 250px;
}
li:hover {
	background-color: rgba(0, 0, 0, 0.05);
}
</style>
