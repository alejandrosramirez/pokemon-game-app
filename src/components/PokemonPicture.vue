<template>
	<div class="pokemon-picture-container">
		<img
			v-if="!showPokemonOption"
			:src="imageSrc"
			class="pokemon-picture-hidden"
			alt="pokemon"
		/>

		<img v-else :src="imageSrc" class="fade-in" alt="pokemon" />
	</div>
</template>

<script>
export default {
	name: "PokemonPicture",
	props: {
		pokemonOptionId: {
			type: Number,
			required: true,
		},
		showPokemonOption: {
			type: Boolean,
			required: true,
			default: false,
		},
	},
	computed: {
		imageSrc() {
			return `https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/${this.pokemonOptionId}.svg`;
		},
	},
};
</script>

<style scoped>
.pokemon-picture-container {
	height: 200px;
}
img {
	height: 200px;
	margin-left: -50vw;
	position: absolute;
	width: 100vw;
	user-select: none;
	-moz-user-select: none;
	-ms-user-select: none;
	-webkit-user-drag: none;
	-webkit-user-select: none;
}
.pokemon-picture-hidden {
	filter: brightness(0);
}
</style>
