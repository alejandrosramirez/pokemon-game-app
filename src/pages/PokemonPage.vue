<template>
	<h1 v-if="!pokemonOption">Espere por favor...</h1>
	<div v-else>
		<h1>¿Quién es ese Pokémon?</h1>

		<PokemonPicture
			:pokemonOptionId="pokemonOption.id"
			:showPokemonOption="showPokemonOption"
		/>

		<PokemonOptions
			:pokemonOptions="pokemonOptions"
			@selectedPokemon="checkPokemonOption"
		/>

		<template v-if="showSuccessPokemonOption">
			<h2>{{ message }}</h2>

			<button @click="newGame">
				Nueva partida
			</button>
		</template>
	</div>
</template>

<script>
import PokemonPicture from "@/components/PokemonPicture.vue";
import PokemonOptions from "@/components/PokemonOptions.vue";

import getPokemonOptions from "@/helpers/getPokemonOptions";

export default {
	name: "PokemonPage",
	components: {
		PokemonPicture,
		PokemonOptions,
	},
	data() {
		return {
			pokemonOptions: [],
			pokemonOption: null,
			showPokemonOption: false,
			showSuccessPokemonOption: false,
			message: "",
		};
	},
	methods: {
		async getPokemonOptionsArr() {
			this.pokemonOptions = await getPokemonOptions();

			const randInt = Math.floor(Math.random() * 4);

			this.pokemonOption = this.pokemonOptions[randInt];
		},
		checkPokemonOption(pokemonOptionId) {
			this.showPokemonOption = true;
			this.showSuccessPokemonOption = true;

			if (pokemonOptionId === this.pokemonOption.id) {
				this.message = `Es... ${this.pokemonOption.name}`;
			} else {
				this.message = `No... es ${this.pokemonOption.name}`;
			}
		},
		async newGame() {
			this.pokemonOptions = [];
			this.pokemonOption = null;
			this.showPokemonOption = false;
			this.showSuccessPokemonOption = false;
			this.message = "";
			await this.getPokemonOptionsArr();
		}
	},
	mounted() {
		this.getPokemonOptionsArr();
	},
};
</script>
