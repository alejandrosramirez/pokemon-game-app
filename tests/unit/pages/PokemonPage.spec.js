import { shallowMount, mount } from "@vue/test-utils";

import PokemonPage from "@/pages/PokemonPage.vue";
import PokemonPicture from "@/components/PokemonPicture.vue";
import PokemonOptions from "@/components/PokemonOptions.vue";
import { pokemonOptions } from "../mocks/pokemonOptions.mock";

describe("Pruebas en <PokemonPage />", () => {
	let wrapper;

	beforeEach(() => {
		wrapper = shallowMount(PokemonPage);
	});

	test("Debe de hacer match con el snapshot", () => {
		expect(wrapper.html()).toMatchSnapshot();
	});

	test("Debe de de llamar 'getPokemonOptionsArr' al montarse", () => {
		const getPokemonOptionsArrSpy = jest.spyOn(
			PokemonPage.methods,
			"getPokemonOptionsArr"
		);
		shallowMount(PokemonPage);
		expect(getPokemonOptionsArrSpy).toHaveBeenCalled();
	});

	test("Debe de hacer match con el snapshot cuando se cargan los pokÃ©mon", () => {
		const wrapper = shallowMount(PokemonPage, {
			data() {
				return {
					pokemonOptions,
					pokemonOption: pokemonOptions[0],
					showPokemonOption: false,
					showSuccessPokemonOption: false,
					message: "",
				};
			},
		});
		expect(wrapper.html()).toMatchSnapshot();
	});

	test("Debe de mostrar los componentes 'PokemonPicture' y PokemonOptions'", () => {
		const wrapper = shallowMount(PokemonPage, {
			data() {
				return {
					pokemonOptions,
					pokemonOption: pokemonOptions[0],
					showPokemonOption: false,
					showSuccessPokemonOption: false,
					message: "",
				};
			},
		});
		const pokemonPictureStub = wrapper.find("pokemon-picture-stub");
		const pokemonOptionsStub = wrapper.find("pokemon-options-stub");
		expect(pokemonPictureStub.exists()).toBeTruthy();
		expect(pokemonOptionsStub.exists()).toBeTruthy();
		expect(pokemonPictureStub.attributes("pokemonoptionid")).toBe("1");
		expect(pokemonOptionsStub.attributes("pokemonoptions")).toBeTruthy();
	});

	test("Debe de probar 'checkPokemonOption'", async () => {
		const wrapper = shallowMount(PokemonPage, {
			data() {
				return {
					pokemonOptions,
					pokemonOption: pokemonOptions[0],
					showPokemonOption: false,
					showSuccessPokemonOption: false,
					message: "",
				};
			},
		});
		await wrapper.vm.checkPokemonOption(1);
		expect(wrapper.find("h2").exists()).toBeTruthy();
		expect(wrapper.vm.showPokemonOption).toBeTruthy();
		expect(wrapper.find("h2").text()).toBe(`Es... ${pokemonOptions[0].name}`);
		await wrapper.vm.checkPokemonOption(3);
		expect(wrapper.vm.message).toBe(`No... es ${pokemonOptions[0].name}`);
	});
});
