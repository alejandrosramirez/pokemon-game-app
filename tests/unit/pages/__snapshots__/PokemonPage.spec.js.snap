// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <PokemonPage /> Debe de hacer match con el snapshot 1`] = `<h1>Espere por favor...</h1>`;

exports[`Pruebas en <PokemonPage /> Debe de hacer match con el snapshot cuando se cargan los pokémon 1`] = `
<div>
  <h1>¿Quién es ese Pokémon?</h1>
  <pokemon-picture-stub pokemonoptionid="1" showpokemonoption="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemonoptions="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
